src/main/java/org/spongepowered/common/event/tracking/PooledPhaseState.java

```diff
@@ -25,45 +25,62 @@
 package org.spongepowered.common.event.tracking;
 
 import java.util.ArrayDeque;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
 
 import javax.annotation.Nullable;
 
 public abstract class PooledPhaseState<C extends PhaseContext<C>> implements IPhaseState<C> {
 
-    private final ArrayDeque<C> contextPool = PhaseTracker.SERVER.createContextPool(this);
+    private static final ConcurrentHashMap<IPhaseState<?>, ArrayDeque<? extends PhaseContext<?>>> stateContextPool = new ConcurrentHashMap<>();
+    private final ArrayDeque<C> contextPool;
     @Nullable private C cached;
+    final ReentrantLock lock = new ReentrantLock();
 
     protected PooledPhaseState() {
+        final ArrayDeque<C> pool = new ArrayDeque<>();
+        PooledPhaseState.stateContextPool.put(this, pool);
+        this.contextPool = pool;
     }
 
     @Override
     public final C createPhaseContext() {
-        if (this.cached != null && !this.cached.isCompleted) {
-            final C cached = this.cached;
-            this.cached = null;
-            return cached;
+        this.lock.lock();
```

Comment:
@dualspiral @Zidane I'm introducing this due to the nature of now having possibly multiple threads accessing the same values/stacks, and avoiding the cache getting corrupted, I figured this would be optimal, but I'd ask for someone more knowledgeable than I about threading.

Reply:
Hooray, my numerical simulation knowledge comes into play here!

This will work, but I'm concerned about the performance penalty of this. Now, I'm not completely versed in the internals of the PhaseTracker, but I know enough about it to know that this code is going to be executed hundereds, if not thousands of times per second.

Consider this happening on multiple threads. A lock must always be treated as a bottleneck, because you're saying that you know multiple threads might access this so you need them to do it one at a time. That's going to happen a lot here - if we're not careful, the performance benefit of the async world generators will become a performance deficit because phases can only be switched one at a time - and this is happening both on build up and tear down of a phase.

I think that we may need to give each instance of the phase tracker a `UUID` that allows us to identify it, and/or let each phase tracker have its own pool of objects. I don't think there will be a massive cost memory wise _really_ and we'll not have to worry about dreadful performance because of many many cross cutting contexts switches a second.

(Aside: I've done this before. I used MPI in my Monte-Carlo simulations but needed everything to sync up at one point. I was too judicious with my sync points and my simulation ended up being three times slower. I also had the risk of deadlock and this affected multiple nodes in a cluster hitting many other users waiting for their code to run! I don't think we have it here but that was hard to resolve...)

Reply:
Looking a slight bit more, the cached context is linked to the state which I'm also guessing is pooled - is the state per phase-tracker instance? If it is, you wouldn't need to lock until after the first if block which would mitigate some of the hit with locking as the cached context would also be per phase-tracker instance.

Reply:
> Hooray, my numerical simulation knowledge comes into play here!

Yay! Primarily why I ask, since 

> Consider this happening on multiple threads. A lock must always be treated as a bottleneck, because you're saying that you know multiple threads might access this so you need them to do it one at a time. That's going to happen a lot here - if we're not careful, the performance benefit of the async world generators will become a performance deficit because phases can only be switched one at a time - and this is happening both on build up and tear down of a phase.

Right, suffice to say that this will happen for each `Entity`, `TileEntity`, `BlockState`, `FluidState`, and `BlockState` random ticks, along with every single `NeighborNotification` and possibly `NeighborShapeUpdate`, so it's going to be "hot" to say the least.

> I think that we may need to give each instance of the phase tracker a `UUID` that allows us to identify it, and/or let each phase tracker have its own pool of objects. I don't think there will be a massive cost memory wise _really_ and we'll not have to worry about dreadful performance because of many many multi thread contexts switches a second.

Currently each `PhaseTracker` gets it's own `Thread` and I've contemplated on generating some `UUID`s, possibly Type5 based on the "thread namespace", thereby guaranteeing the same uuid constant for the server thread and client thread, but world gen threads I've yet to find where those get spun off to hook a new PhaseTracker instance to them. I made this change to avoid the pool of objects existing on the `PhaseTracker` since it did nothing with the pools except shove them to a map for debugging purposes. If we are to move to per-`PhaseTracker` context creation, then we'll also need to supply the `PhaseTracker` instance being used (which is based on thread and expectations).

> (Aside: I've done this before. I used MPI in my Monte-Carlo simulations but needed everything to sync up at one point. I was too judicious with my sync points and my simulation ended up being three times slower. I also had the risk of deadlock and this affected multiple nodes in a cluster hitting many other users waiting for their code to run! I don't think we have it here but that was hard to resolve...)

I do understand the possibilities of deadlocking (There's a relatively easy deadlock you can do with a `Feature` that spawns entities and if any methods in the `Entity` call to the parent world to get a `Chunk`, boom, instant deadlock.

> Looking a slight bit more, the cached context is linked to the state which I'm also guessing is pooled - is the state per phase-tracker instance? If it is, you wouldn't need to lock until after the first if block which would mitigate some of the hit with locking as the cached context would also be per phase-tracker instance.

Currently every state instance is considered a singleton. We don't have pooled states since the `IPhaseState` is effectively a controller while the `PhaseContext` is the model. It's possible to shove the context pooling to the `PhaseTracker` instance since the context will know which is being used when it's being "switched" in.